// @cnpja/sdk/dist/cnpja/cnpja.module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Cnpja = void 0;
const ccc_service_1 = require("../ccc/ccc.service");
const company_service_1 = require("../company/company.service");
const credit_service_1 = require("../credit/credit.service");
const http_service_1 = require("../http/http.service");
const list_service_1 = require("../list/list.service");
const office_service_1 = require("../office/office.service");
const person_service_1 = require("../person/person.service");
const rfb_service_1 = require("../rfb/rfb.service");
const simples_service_1 = require("../simples/simples.service");
const suframa_service_1 = require("../suframa/suframa.service");
const zip_service_1 = require("../zip/zip.service");
class Cnpja {
    constructor(options) {
        this.options = options;
        this.setup();
    }
    /**
     * Resolve dependency tree and inject providers.
     */
    setup() {
        const { apiKey, baseUrl, headers, retryLimit, retryDelay } = this.options;
        const httpService = new http_service_1.HttpService({
            baseUrl: baseUrl || 'https://api.cnpja.com',
            headers: { ...(headers || {}), Authorization: apiKey },
            retryLimit: retryLimit ?? 3,
            retryDelay: retryDelay || ((r) => 2 ** (r - 1) * 1000),
        });
        this.office = new office_service_1.OfficeService(httpService);
        this.company = new company_service_1.CompanyService(httpService);
        this.person = new person_service_1.PersonService(httpService);
        this.rfb = new rfb_service_1.RfbService(httpService);
        this.simples = new simples_service_1.SimplesService(httpService);
        this.ccc = new ccc_service_1.CccService(httpService);
        this.suframa = new suframa_service_1.SuframaService(httpService);
        this.zip = new zip_service_1.ZipService(httpService);
        this.list = new list_service_1.ListService(httpService);
        this.credit = new credit_service_1.CreditService(httpService);
    }
}
exports.Cnpja = Cnpja;
