// @cnpja/sdk/dist/list/list.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListService = void 0;
class ListService {
    constructor(httpService) {
        this.httpService = httpService;
    }
    /**
     * ### Cria Lista​
     * Adiciona uma nova lista em sua conta, é possível criar até mil listas.
     * @param params
     */
    create(params) {
        return this.httpService.post('list', {
            json: params,
        });
    }
    /**
     * ### Pesquisa Listas​
     * Encontra as listas que correspondem aos filtros configurados.
     * @param params
     */
    async *search(params) {
        yield* this.httpService.getPage('list', {
            query: params,
        });
    }
    /**
     * ### Consulta Lista
     * Adquire informações de uma lista através de seu identificador único.
     * @param params
     */
    async read(params) {
        const { listId } = params;
        return this.httpService.get('list/:listId', {
            replacements: { listId },
        });
    }
    /**
     * ### Atualiza Lista​
     * Modifica as informações de uma lista através de seu identificador único.
     * @param params
     */
    async update(params) {
        const { listId, ...json } = params;
        return this.httpService.patch('list/:listId', {
            replacements: { listId },
            json,
        });
    }
    /**
     * ### Remove Lista
     * Apaga uma lista através de seu identificador único.
     * @param params
     */
    async delete(params) {
        const { listId } = params;
        return this.httpService.delete('list/:listId', {
            replacements: { listId },
        });
    }
    /**
     * ### Exporta Lista​
     * Cria uma tarefa de exportação da lista em formatos ZIP e Excel, obedecendo aos parâmetros de consulta CNPJ.
     * @param params
     */
    createExport(params) {
        const { listId, ...json } = params;
        return this.httpService.post('list/:listId/export', {
            replacements: { listId },
            json,
        });
    }
    /**
     * ### Pesquisa Exportações
     * Encontra as exportações de uma lista através de seu identificador único.
     * @param params
     */
    async *searchExport(params) {
        const { listId, ...query } = params;
        yield* this.httpService.getPage('list/:listId/export', {
            replacements: { listId },
            query,
        });
    }
    /**
     * ### Consulta Exportação​
     * Adquire as informações da exportação de uma lista através de seus identificadores únicos.
     * @param params
     */
    readExport(params) {
        return this.httpService.post('list/:listId/export/:exportId', {
            replacements: params,
        });
    }
}
exports.ListService = ListService;
