// @cnpja/sdk/dist/office/office.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OfficeService = void 0;
class OfficeService {
    constructor(httpService) {
        this.httpService = httpService;
    }
    /**
     * ### Consulta CNPJ
     * Adquire de forma centralizada múltiplas informações de um estabelecimento,
     * incluindo Receita Federal, Simples Nacional e Cadastro de Contribuintes.
     * @param params
     */
    read(params) {
        const { taxId, ...query } = params;
        return this.httpService.get('office/:taxId', {
            replacements: { taxId },
            query,
        });
    }
    /**
     * ### Pesquisa CNPJ
     * Lista todos os estabelecimentos que correspondem aos filtros configurados.
     * @param params
     */
    async *search(params) {
        yield* this.httpService.getPage('office', {
            query: params,
        });
    }
    /**
     * ### Mapa Aéreo
     * Gera o mapa aéreo referente ao endereço do estabelecimento.
     * @param params
     */
    map(params) {
        const { taxId, ...query } = params;
        return this.httpService.get('office/:taxId/map', {
            replacements: { taxId },
            query,
        });
    }
    /**
     * ### Visão da Rua
     * Gera a visão da rua referente ao endereço do estabelecimento.
     * @param params
     */
    street(params) {
        const { taxId, ...query } = params;
        return this.httpService.get('office/:taxId/street', {
            replacements: { taxId },
            query,
        });
    }
}
exports.OfficeService = OfficeService;
